src/wrapper/objects/global_ref.rs

```diff
@@ -165,25 +170,19 @@ impl Drop for GlobalRefGuard {
     fn drop(&mut self) {
         let raw: sys::jobject = mem::take(&mut self.obj).into_raw();
 
-        let drop_impl = |env: &JNIEnv| -> Result<()> {
+        let res: Result<()> = (|| {
+            let env = self.vm.attach_current_thread()?;
```

Comment:
My slight reservation here is that I'm half expecting that when we look more closely at the safety of the attachment APIs, we're going to conclude that we should remove the RAII/guard based attachments, at least until we can create a proper singleton crate that can guarantee we only have a single thread-local for tracking the attachments. (currently multiple versions of the jni crate or mixing with alternative Rust jni crates could lead to unreliable attachment checks)

The end result is more readable here but it looks like the previous code has the same behaviour effectively, except it relies on the jni library to do the thread-local look up. Is there maybe a significant performance difference here?

Reply:
Yes, it has the same behavior as before.

As you know, the `GlobalRef` destructor has to delete the global reference on a potentially unattached thread, so it tries attaching the current thread and, if the current thread is not already attached, temporarily attaches it and logs a warning.

Getting a Java exception message from a `Display` implementation has the same problem, so I refactored the attach-with-warning code into its own function, which is called from the `GlobalRef` destructor as well as from `impl Display for JavaException` [here](https://github.com/jni-rs/jni-rs/blob/138006d14a8cac865b4a8d40130b03e8e3422fde/src/wrapper/errors/exception_capture.rs#L244).

> I'm half expecting that when we look more closely at the safety of the attachment APIs, we're going to conclude that we should remove the RAII/guard based attachments, at least until we can create a proper singleton crate that can guarantee we only have a single thread-local for tracking the attachments.

Probably. The safety issue I'm aware of is that two `AttachGuards` could be acquired at the same time, and then the first one could be dropped before the second one is dropped, resulting in an invalid `JNIEnv`. This could be solved with a thread-local variable that counts how many times the thread has been attached, so that it is only actually detached when the counter drops to zero, but yeah, there has to be exactly one such counter that everyone uses. Pity the JVM doesn't do that…

I don't think all of that applies here, though, because this function only acquires a single `AttachGuard`, and doesn't call anything that could potentially acquire a second one.